share/functions/umask.fish

```diff
@@ -1,117 +1,124 @@
+    if not string match -qr "^$basic_pattern(,$basic_pattern)*\$" -- $argv
+        printf (_ "%s: Invalid mask '%s'\n") umask $argv >&2
+        return 1
+    end
 
-            if not echo $perm | __fish_sgrep -E '^(r|w|x)*$' >/dev/null
-                printf (_ "%s: Invalid mask '%s'\n") umask $argv >&2
-                return
-            end
+    # Insert inverted umask into res variable.
+    set -l tmp (string split '' $umask)
+    set -l res
+    set res[1] (math 7 - $tmp[2])
+    set res[2] (math 7 - $tmp[3])
+    set res[3] (math 7 - $tmp[4])
+
+    for rights in (string split , $argv)
+        set -l match (string match -r "^$basic_pattern\$" $rights)
+        set -l scope $match[2]
+        set -l mode $match[3]
+        set -l perms $match[4]
+        if test -n "$scope" -a -z "$mode"
+            printf (_ "%s: Invalid mask '%s'\n") umask $argv >&2
+            return 1
+        end
+        if test -z "$scope"
+            set scope a
+        end
+        if test -z "$mode"
+            set mode =
+        end
 
-            set val 0
-            if echo $i | __fish_sgrep 'r' >/dev/null
-                set val 4
-            end
-            if echo $i | __fish_sgrep 'w' >/dev/null
-                set val (math $val + 2)
-            end
-            if echo $i | __fish_sgrep 'x' >/dev/null
-                set val (math $val + 1)
-            end
+        set -l scopes_to_modify
+        string match -q '*u*' $scope
+        and set scopes_to_modify 1
+        string match -q '*g*' $scope
+        and set scopes_to_modify $scopes_to_modify 2
+        string match -q '*o*' $scope
+        and set scopes_to_modify $scopes_to_modify 3
+        string match -q '*a*' $scope
+        and set scopes_to_modify 1 2 3
+
+        set -l val 0
+        if string match -q '*r*' $perms
+            set val 4
+        end
+        if string match -q '*w*' $perms
+            set val (math $val + 2)
+        end
+        if string match -q '*x*' $perms
+            set val (math $val + 1)
+        end
 
-            for j in $idx
-                switch $mode
-                    case set
-                        set res[$j] $val
+        for j in $scopes_to_modify
+            switch $mode
+                case '='
+                    set res[$j] $val
 
-                    case add
-                        set res[$j] (perl -e 'print( ( '$res[$j]'|'$val[$j]' )."\n" )')
+                case '+'
+                    set res[$j] (__fish_umask_add $res[$j] $val)
 
-                    case remove
-                        set res[$j] (perl -e 'print( ( (7-'$res[$j]')&'$val[$j]' )."\n" )')
-                end
+                case '-'
+                    set res[$j] (__fish_umask_remove $res[$j] $val)
             end
         end
+    end
 
-        for i in 1 2 3
-            set res[$i] (math 7 - $res[$i])
-        end
-        echo 0$res[1]$res[2]$res[3]
+    for i in 1 2 3
+        set res[$i] (math 7 - $res[$i])
```

Comment:
Correct. If I change the code to eliminate the earlier math calls then this can also be eliminated.

Thanks for taking the time to look at this PR and provide feedback.